### PR TITLE
Introduce symbol versioning

### PR DESCRIPTION
**Pull request by ansasaki**
This adds a linker map, which adds version information for exported
symbols, and the option to compile with symbol versioning. After adding
symbol versioning, modifications in the exported interface must be
treated with care and the linker map updated accordingly.

### CODE REVIEW

File: libssh.map [hunk 1]
```diff
@@ -0,0 +1,425 @@
+# When new symbols are added in a release, they must be added as:
```
Comment (ueno) [inline libssh.map]: nit: it might be better to place this file under `src/` as it is only referred to from `src/CMakeList.txt`?

Comment (ansasaki) [reply libssh.map]: Yes, it makes sense to store it in ``src/``. My only concern is about future maintenance of the linker map that have to be done in each release. Storing it in ``src/`` can make the maintainer to forget to do it. I think I should add rules in the ``Makefile`` to check the ABI (using ``abi-compliance-checker``)

**Comment by ansasaki** [comment]
Thanks for the review!
I will check the availability of the ``-version-script`` option in ``ConfigureChecks.cmake``. If it is available, I'll replace the option.

**Comment by ansasaki** [comment]
I moved the linker map to ``src/`` and made the option to build with symbol versioning enabled by default. The option is disabled in non UNIX like platforms.

**Pull request merged**